fsm.go

```diff
@@ -25,221 +25,168 @@
 package fsm
 
 import (
-	"strings"
+	"fmt"
 	"sync"
+
+	"golang.org/x/exp/constraints"
 )
 
 // transitioner is an interface for the FSM's transition function.
-type transitioner interface {
-	transition(*FSM) error
+type transitioner[E constraints.Ordered, S constraints.Ordered] interface {
+	transition(*FSM[E, S]) error
 }
 
 // FSM is the state machine that holds the current state.
-//
-// It has to be created with NewFSM to function properly.
-type FSM struct {
+// E ist the event
+// S is the state
+// It has to be created with New to function properly.
+type FSM[E constraints.Ordered, S constraints.Ordered] struct {
 	// current is the state that the FSM is currently in.
-	current string
+	current S
 
 	// transitions maps events and source states to destination states.
-	transitions map[eKey]string
+	transitions map[eKey[E, S]]S
 
 	// callbacks maps events and targets to callback functions.
-	callbacks map[cKey]Callback
+	callbacks Callbacks[E, S]
 
 	// transition is the internal transition functions used either directly
 	// or when Transition is called in an asynchronous state transition.
 	transition func()
-	// transitionerObj calls the FSM's transition() function.
-	transitionerObj transitioner
+	// transitioner calls the FSM's transition() function.
+	transitioner transitioner[E, S]
 
 	// stateMu guards access to the current state.
 	stateMu sync.RWMutex
 	// eventMu guards access to Event() and Transition().
 	eventMu sync.Mutex
 	// metadata can be used to store and load data that maybe used across events
 	// use methods SetMetadata() and Metadata() to store and load data
-	metadata map[string]interface{}
+	metadata map[string]any
 
 	metadataMu sync.RWMutex
 }
 
-// EventDesc represents an event when initializing the FSM.
+// Transition represents an event when initializing the FSM.
 //
 // The event can have one or more source states that is valid for performing
 // the transition. If the FSM is in one of the source states it will end up in
 // the specified destination state, calling all defined callbacks as it goes.
-type EventDesc struct {
-	// Name is the event name used when calling for a transition.
-	Name string
+type Transition[E constraints.Ordered, S constraints.Ordered] struct {
```

Comment:
Why was this renamed? I may like it, but would like to hear your reasoning.

Reply:
Well because in a fsm the mapping between events, source states and destination state is usually called a transition table, see [wikipedia](https://en.wikipedia.org/wiki/Finite-state_machine) for example

Reply:
Sounds reasonable.